rust/src/coordinator_async/message_processing.rs

```diff
@@ -0,0 +1,278 @@
+        } = self;
+        // Collect the results of the MessageHandler tasks. The collect future will be
+        // successfully resolved when the minimum duration has been waited and when the minimum
+        // number of successful validated messages has been reached.
+        // The first failed MessageHandler result causes the collection to be canceled. In this
+        // case the collect future will be resolved with an error.
+        let min_collection_duration = async move {
+            tokio::time::delay_for(min_duration).await;
+            println!("waited min collection duration");
+            Ok::<(), PetError>(())
+        };
+
+        let message_collection = async move {
+            loop {
+                let _message = sink_rx
+                    .recv()
+                    .await
+                    // First '?' operator is applied on the result of '.recv()', the second one on
+                    // the message itself (which has the type 'Result<(), PetError>').
+                    .ok_or(PetError::InvalidMessage)??;
+
+                successful_messages += 1;
+
+                if successful_messages == min_messages {
+                    break Ok::<(), PetError>(());
+                }
+            }
+        };
+
+        let mut max_collection_duration = tokio::time::delay_for(max_duration);
+
+        tokio::select! {
+            collection_result = async {
+                tokio::try_join!(min_collection_duration, message_collection).map(|_| ())
+            }=> {
+                collection_result
+            }
+            _ = &mut max_collection_duration => {
+                println!("message collection timed out!");
+                Err::<(), PetError>(PetError::InvalidMessage)
+            }
+        }
+    }
+}
+
+pub struct MessageHandler {
+    // A sender through which the result of the massage validation is sent.
+    sink_tx: UnboundedSender<Result<(), PetError>>,
+    // We will never send anything on this channel.
+    // We use this channel to keep track of which MessageHandler tasks are still alive.
+    _cancel_complete_tx: Sender<()>,
+    // The channel receiver that receives the cancel notification.
+    notify_cancel: broadcast::Receiver<()>,
+}
+
+impl MessageHandler {
+    pub fn new(
+        sink_tx: UnboundedSender<Result<(), PetError>>,
+        cancel_complete_tx: Sender<()>,
+        notify_cancel: broadcast::Receiver<()>,
+    ) -> Self {
+        Self {
+            sink_tx,
+            _cancel_complete_tx: cancel_complete_tx,
+            notify_cancel,
+        }
+    }
+
+    async fn handle_message(
+        self,
+        message_validation_fut: impl Future<Output = Result<(), PetError>>,
+    ) {
+        // Extract all fields of the MessageHandler struct. This is necessary to bypass borrow
+        // issues in the tokio::select macro.
+        // It is important to extract _cancel_complete_tx as well, otherwise the channel
+        // will be dropped too early.
```

Comment:
this issue might be avoided by not destructuring `self`, but instead use `self.sink_tx` and `self.notify_cancel` in the `select!` macro. then `self` will be dropped at the end of the method. this would require `mut self` instead of `self` as an argument to `handle_message()`.